### PR TITLE
RQM fixes for custom Attributes

### PR DESCRIPTION
**Pull request by prashelke**
Thank you for your contribution.
Before submitting this PR, please make sure:

- [x] PR description and commit message should describe the changes done in this PR
- [x] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [x] Latest Code from master or specific release branch is merged to your branch
- [x] No unwanted\commented\junk code is included
- [ ] No new warning upon build solution
- [x] Code Summary\Comments are added to my code which explains what my code is doing
- [ ] Existing unit test cases are passed
- [ ] New Unit tests are added for your development
- [ ] Sanity Tests are successfully executed for New and Existing Functionality
- [ ] Verify that changes are compatible with all relevant browsers and platforms.
- [x] After creating pull request there should not be any conflicts
- [ ] Resolve all Codacy comments
- [ ] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Update the Help Library document to match any feature changes

## Summary by CodeRabbit

- **New Features**
	- Enhanced functionality for exporting execution details to RQM, including validation of external fields and improved user feedback.
	- Improved error handling and performance optimizations in the importation of RQM test plans, including better management of test cases and parameters.
	- Added functionality for mapping external fields with enhanced data handling and validation.

- **Bug Fixes**
	- Resolved issues with null test plans and enhanced logging for importing test cases and parameters.

- **Documentation**
	- Updated method signatures to reflect new parameters and improved functionality across multiple classes.

### CODE REVIEW

File: Ginger/GingerCoreNET/ALMLib/RQM/ImportFromRQM.cs [hunk 1]
```diff
@@ -1364,14 +1361,32 @@ private static ObservableList<ExternalItemFieldBase> GetCustomAttributes(Backgro
...
+                                XmlNode requiredNode = CustomAttributeListing.SelectSingleNode(xpath, nsManager);
+
+                                if (requiredNode != null)
+                                {
+                                    CustomAttributeMandatory = CustomAttributeListing.GetElementsByTagName("ns2:required").Item(0).InnerText;
+                                }
+                                else
+                                {
+                                    CustomAttributeMandatory = "false";
+                                }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/ALMLib/RQM/ImportFromRQM.cs:1381]: Done

File: Ginger/GingerCoreNET/ALMLib/RQM/ImportFromRQM.cs [hunk 2]
```diff
@@ -1442,28 +1457,59 @@ private static ObservableList<ExternalItemFieldBase> GetCustomAttributes(Backgro
...
+                            XmlNode requiredNode = CustomAttributeListing.SelectSingleNode(xpath, nsManager);
+
+                            if (requiredNode != null)
+                            {
+                                CustomAttributeMandatory = CustomAttributeListing.GetElementsByTagName("ns2:required").Item(0).InnerText;
+                            }
+                            else
+                            {
+                                CustomAttributeMandatory = "false";
+                            }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/ALMLib/RQM/ImportFromRQM.cs]: done

File: Ginger/GingerCoreNET/ALMLib/RQM/ExportToRQM.cs [hunk 3]
```diff
@@ -925,9 +925,16 @@ public bool ExportBusinessFlowToRQM(BusinessFlow businessFlow, ObservableList<Ex
...
                     field.SelectedValue = fields[indx].SelectedValue;
+                    field.ID = fields[indx].ID;
+                    field.TypeIdentifier = fields[indx].TypeIdentifier;
+                    field.IsMultiple = fields[indx].IsMultiple;
+                    
+                if (!fieldsToReturn.Contains(field))
+                {
+                    // Add it if it doesn't already exist
                     fieldsToReturn.Add(field);
+                }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/ALMLib/RQM/ExportToRQM.cs]: Done

File: Ginger/GingerCoreNET/ALMLib/RQM/ImportFromRQM.cs [hunk 4]
```diff
@@ -1442,28 +1481,60 @@ private static ObservableList<ExternalItemFieldBase> GetCustomAttributes(Backgro
...
+
+                            if (requiredNode != null)
+                            {
+                                CustomAttributeMandatory = requiredNode.InnerText;
+                            }
+                            else
+                            {
+                                CustomAttributeMandatory = "false";
+                            }
+
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/ALMLib/RQM/ImportFromRQM.cs:1502]: not needed

**Review by Maheshkale447 — APPROVED** [review]

File: Ginger/GingerCoreNET/GeneralLib/General.cs [hunk 5]
```diff
@@ -740,19 +744,68 @@ public static ObservableList<ExternalItemFieldBase> GetExternalFields()
...
-                IsCustomField = externalItemField.IsCustomField,
-                SelectedValue = existingField != null && !string.IsNullOrEmpty(existingField.SelectedValue)
-                    ? existingField.SelectedValue
-                    : externalItemField.SelectedValue
+                "INTEGER" => "1",
+                "MEDIUMSTRING" => "Dummy",
+                "SMALLSTRING" => "Dummy",
+                _ => externalItemField.SelectedValue
             };
         }
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/GeneralLib/General.cs:810]: no need

File: Ginger/GingerCoreNET/GeneralLib/General.cs [hunk 6]
```diff
@@ -740,19 +744,68 @@ public static ObservableList<ExternalItemFieldBase> GetExternalFields()
...
+                return new ExternalItemFieldBase
+                {
+                    Name = externalItemField.Name,
+                    ID = externalItemField.ID,
+                    ItemType = externalItemField.ItemType,
+                    Type = externalItemField.Type,
+                    Guid = externalItemField.Guid,
+                    IsCustomField = externalItemField.IsCustomField,
+                    SelectedValue = value
+                };
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/GeneralLib/General.cs:791]: no  need

File: Ginger/GingerCoreNET/GeneralLib/General.cs [hunk 7]
```diff
@@ -740,19 +744,68 @@ public static ObservableList<ExternalItemFieldBase> GetExternalFields()
...
+            }
+            catch (Exception ex)
+            {
+                Reporter.ToLog(eLogLevel.ERROR,"Failed to Map External Fields",ex.InnerException);
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/GingerCoreNET/GeneralLib/General.cs:795]: will handle it in next release

**Pull request merged**